### PR TITLE
implemented isposinf

### PR DESCRIPTION
**Pull request by zuhorer**
## Proposed changes

Please include a description of the problem or feature this PR is addressing. If there is a corresponding issue, include the issue #.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/ops.cpp [hunk 1]
```diff
@@ -1053,6 +1053,13 @@ array array_equal(
 array isnan(const array& a, StreamOrDevice s /* = {} */) {
   return not_equal(a, a, s);
 }
+array isposinf(const array& a, StreamOrDevice s) {
+  array pos_inf_array =
+      full(a.shape(), std::numeric_limits<double>::infinity());
+
+  array result = a == pos_inf_array;
+  return result;
```
Comment (awni) [inline mlx/ops.cpp:1061]:
```suggestion
  return equal(a, array(std::numeric_limits<float>::infinity(), a.dtype()), s)
```

Make sure you use the stream. Also the equal will broadcast appropriately.

File: python/src/ops.cpp [hunk 2]
```diff
@@ -1838,6 +1838,22 @@ void init_ops(py::module_& m) {
...
+      py::kw_only(),
+      "stream"_a = none,
+      R"pbdoc(
+        isposinf(a: array, stream: Union[None, Stream, Device] = None) -> array
+        Return a boolean array indicating which elements are positive infinity.
+        Args:
+            a (array): Input array.
+            stream (Union[None, Stream, Device]): Optional stream or device.
+        Returns:
+            array: The array with boolean values indicating which elements are positive infinity.
```
Comment (awni) [inline python/src/ops.cpp:1855]: Please add returns in the docstring (following the style of other docstrings in this file).

File: python/tests/test_ops.py [hunk 3]
```diff
@@ -336,6 +336,22 @@ def test_isnan(self):
...
+        x = mx.array([0.0, float("inf")]).astype(mx.float16)
+        self.assertEqual(mx.isposinf(x).tolist(), [False, True])
+
+        x = mx.array([0.0, float("inf")]).astype(mx.bfloat16)
+        self.assertEqual(mx.isposinf(x).tolist(), [False, True])
+
+        x = mx.array([0.0, float("inf")]).astype(mx.complex64)
+        self.assertEqual(mx.isposinf(x).tolist(), [False, True])
+
+        self.assertEqual(mx.isposinf(0 * mx.array(float("-inf"))).tolist(), False)
```
Comment (awni) [inline python/tests/test_ops.py:352]: Could you test with just `float("-inf")` instead of `NaN` here?

File: python/src/array.cpp [hunk 4]
```diff
@@ -1132,6 +1132,12 @@ void init_array(py::module_& m) {
           py::kw_only(),
           "stream"_a = none,
           "See :func:`any`.")
+      .def(
+          "isposinf",
+          &mlx::core::isposinf,
```
Comment (awni) [inline python/src/array.cpp:1137]: I don't think we need this method (numpy doesn't have it and I don't see a real need).

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for the contribution, I left a few comments!

File: python/src/ops.cpp [hunk 5]
```diff
@@ -1838,6 +1838,22 @@ void init_ops(py::module_& m) {
         Returns:
             array: The array with boolean values indicating which elements are NaN.
       )pbdoc");
+   m.def(
+      "isposinf",
+      &mlx::core::isposinf,
```
Comment (awni) [inline python/src/ops.cpp:1843]: You don't need the namespace here..

File: python/src/ops.cpp (hunk 2 above)
Comment (zuhorer) [reply python/src/ops.cpp:1855]: reformatted the docstrings

File: python/src/array.cpp (hunk 4 above)
Comment (zuhorer) [reply python/src/array.cpp:1137]: so we do not need to expose this from python right?

File: python/src/ops.cpp (hunk 5 above)
Comment (zuhorer) [reply python/src/ops.cpp]: removed the namespace

**Comment by zuhorer** [comment]
Hi @awni I updated the required changes in this PR, could please review, if this is correct, I will replicate the same on the isneginf PR as well! Thanks for the review!

File: python/src/ops.cpp [hunk 6]
```diff
@@ -1838,6 +1838,25 @@ void init_ops(py::module_& m) {
...
+        isposinf(a: array, stream: Union[None, Stream, Device] = None) -> array
+
+        Return a boolean array indicating which elements are positive infinity.
+
+        Args:
+            a (array): Input array.
+            stream (Union[None, Stream, Device]): Optional stream or device.
+        
+        Returns:
+            array: The array with boolean values indicating which elements are positive infinity.
```
Comment (awni) [inline python/src/ops.cpp:1858]: Nit: `The boolean array indicating ...`

**Review by awni — APPROVED** [review]
Very nice!!

**Comment by awni** [comment]
Also can you double check the `pre-commit` hooks and make sure they ran? (See CONTRIBUTING.md)

**Comment by awni** [comment]
After formatting we can merge, thanks for the contribution!

**Comment by zuhorer** [comment]
Hi @awni I have run the precommit hooks and updated the docstrings!

**Review by awni — CHANGES_REQUESTED** [review]
It looks good, but the code does not compile and the tests do not pass when I fix the compiler error. Please check it and let me know when it's ready for review

**Comment by zuhorer** [comment]
hi @awni I checked this code, it seems there is an issue with this:

array isposinf(const array& a, StreamOrDevice s){return equal(
    a,
    array(std::numeric_limits<float>::infinity(), a.dtype()),
    s)}

**Comment by zuhorer** [comment]
Hi @awni I have implemented both isposinf and isneginf in one PR and ran the precommit, build and tested and raised a new PR.  #470
Could you please review?

**Pull request closed without merging**